gen/generate.go

```diff
@@ -21,60 +21,160 @@
+	ThriftRoot string
+
+	// NoRecurse determines whether code should be generated for included Thrift
+	// files as well. If true, code gets generated only for the first module.
+	NoRecurse bool
 }
 
 // Generate generates code based on the given options.
 func Generate(m *compile.Module, o *Options) error {
-	g := NewGenerator(o.PackageName)
+	if !filepath.IsAbs(o.ThriftRoot) {
+		return fmt.Errorf(
+			"ThriftRoot must be an absolute path: %q is not absolute",
+			o.ThriftRoot)
+	}
+
+	if !filepath.IsAbs(o.OutputDir) {
+		return fmt.Errorf(
+			"OutputDir must be an absolute path: %q is not absolute",
+			o.OutputDir)
+	}
+
+	if o.NoRecurse {
+		return generateModule(m, o)
+	}
 
-	for _, constantName := range constantNames(m.Constants) {
-		c := m.Constants[constantName]
-		if err := Constant(g, c); err != nil {
-			return err
+	return m.Walk(func(m *compile.Module) error {
+		if err := generateModule(m, o); err != nil {
+			return generateError{Name: m.ThriftPath, Reason: err}
 		}
+		return nil
+	})
+}
+
+// generates code for only the given module, assuming that code for included
+// modules has already been generated.
+func generateModule(m *compile.Module, o *Options) error {
+	packagePath, err := filepath.Rel(o.ThriftRoot, m.ThriftPath)
+	if err != nil {
+		return err
 	}
 
-	for _, typeName := range typeNames(m.Types) {
-		t := m.Types[typeName]
-		if err := TypeDefinition(g, t); err != nil {
-			return err
+	// packagePath is the path relative to outputDir into which we'll be
+	// writing the package for this Thrift file. For
+	// $thriftRoot/foo/bar.thrift, packagePath is foo/bar, and packageDir is
+	// $outputDir/foo/bar. All files for bar.thrift will be written to
+	// $outputDir/foo/bar/. The package will be importable via
+	// $packagePrefix/foo/bar.
+	packagePath = strings.TrimSuffix(packagePath, ".thrift")
+	packageDir := filepath.Join(o.OutputDir, packagePath)
+
+	// files contains a collection of files relative to packageDir and their
+	// contents.
+	files := make(map[string]*bytes.Buffer)
+
+	if len(m.Constants) > 0 {
+		g := NewGenerator(o.PackagePrefix, o.ThriftRoot, m.ThriftPath)
+
+		for _, constantName := range sortStringKeys(m.Constants) {
+			if err := Constant(g, m.Constants[constantName]); err != nil {
+				return err
+			}
+		}
+
+		buff := new(bytes.Buffer)
+		if err := g.Write(buff, token.NewFileSet()); err != nil {
+			return fmt.Errorf(
+				"could not generate constants for %q: %v", m.ThriftPath, err)
 		}
+
+		// TODO(abg): Verify no file collisions
+		files["constants.go"] = buff
 	}
 
-	if err := g.Write(o.Output, token.NewFileSet()); err != nil {
-		return err
+	if len(m.Types) > 0 {
+		g := NewGenerator(o.PackagePrefix, o.ThriftRoot, m.ThriftPath)
+
+		for _, typeName := range sortStringKeys(m.Types) {
+			if err := TypeDefinition(g, m.Types[typeName]); err != nil {
+				return err
+			}
+		}
+
+		buff := new(bytes.Buffer)
+		if err := g.Write(buff, token.NewFileSet()); err != nil {
```

Comment:
(not related to this diff) do we use the `token.FileSet` or do we expect to use it from outside of the generator?